compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -1373,406 +1412,323 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
-        val allParamDefsAsVars = jsConstructorBuilder.getAllParamDefsAsVars
-        val overrideNumDef = js.VarDef(overloadIdent, NoOriginalName,
-            jstpe.IntType, mutable = true, js.IntLiteral(0))
+        val assignments = for {
+          (param, arg) <- nextParams.zip(tree.ctor.ctorArgs)
+          if !isUndefinedParam(arg)
+        } yield {
+          js.Assign(genVarRef(param), arg)
+        }
 
-        overrideNumDef :: allParamDefsAsVars ::: newDispatchResolution :: Nil
+        ifOverload(tree, js.Block(inner ++ tree.ctor.beforeCall ++ assignments))
       }
-    }
-
-    private def mkJSConstructorBuilder(ctors: List[js.MethodDef])(
-        implicit pos: Position): JSConstructorBuilder = {
-      def findCtorForwarderCall(tree: js.Tree): MethodName = tree match {
-        case js.ApplyStatic(_, _, method, js.This() :: _)
-            if method.name.isConstructor =>
-          method.name
 
-        case js.Block(stats) =>
-          stats.collectFirst {
-            case js.ApplyStatic(_, _, method, js.This() :: _)
-                if method.name.isConstructor =>
-              method.name
-          }.get
-
-        case _ =>
-          abort(s"Unexpected secondary constructor body at ${tree.pos}:\n$tree")
+      def postStats(tree: ConstructorTree): js.Tree = {
+        val inner = tree.subCtors.map(postStats(_))
+        ifOverload(tree, js.Block(tree.ctor.afterCall ++ inner))
       }
 
-      val (primaryCtor :: Nil, secondaryCtors) = ctors.partition {
-        _.body.get match {
-          case js.Block(stats) =>
-            stats.exists(_.isInstanceOf[js.JSSuperConstructorCall])
-
-          case _: js.JSSuperConstructorCall => true
-          case _                            => false
-        }
-      }
+      js.Block(
+          ctorTrees.map(preStats(_, primaryCtor.params)) ++
+          primaryCtor.body ++
+          ctorTrees.map(postStats(_))
+      )
+    }
 
-      val ctorToChildren = secondaryCtors.map { ctor =>
-        findCtorForwarderCall(ctor.body.get) -> ctor
-      }.groupBy(_._1).map(kv => kv._1 -> kv._2.map(_._2)).withDefaultValue(Nil)
+    private class PrimaryJSCtor(val sym: Symbol,
+        val params: List[Symbol], val body: List[js.Tree])
 
-      var overrideNum = -1
-      def mkConstructorTree(method: js.MethodDef): ConstructorTree = {
-        val subCtrTrees = ctorToChildren(method.methodName).map(mkConstructorTree)
-        overrideNum += 1
-        new ConstructorTree(overrideNum, method, subCtrTrees)
-      }
+    private class SplitSecondaryJSCtor(val sym: Symbol,
+        val params: List[Symbol], val beforeCall: List[js.Tree],
+        val targetCtor: Symbol, val ctorArgs: List[js.Tree],
+        val afterCall: List[js.Tree])
 
-      new JSConstructorBuilder(mkConstructorTree(primaryCtor))
-    }
+    private class ConstructorTree(val lowestNum: Int, val overloadNum: Int,
+        val ctor: SplitSecondaryJSCtor, val subCtors: List[ConstructorTree])
```

Comment:
Have you considered parameterizing this class in the type of `ctor`, so that it can be used for the primary constructor as well? This way we would have a single tree rooted in the primary constructor, instead of a primary constructor on the one hand, and a forest of secondary constructor trees on the other hand. It seems that a lot of code in `genJSClassCapturesAndConstructor` could be simplified that way.

I guess you would need a
```scala
sealed trait JSCtor {
  val sym: Symbol
  val params: List[Symbol]
}
```
that is extended by both `PrimaryJSCtor` and `SplitSecondaryJSCtor`, and then have here:
```suggestion
    private class ConstructorTree[Ctor <: JSCtor](val lowestNum: Int, val overloadNum: Int,
        val ctor: Ctor, val subCtors: List[ConstructorTree[SplitSecondaryJSCtor]])
```

Reply:
I have added another commit with that. More types, less code :P Either is fine with me.

Reply:
I like the new version better 👍 It might not be a *lot* less code, but IMO it is much clearer now.